### PR TITLE
Improve test client

### PR DESCRIPTION
**Pull request by cjellick**
- Fix a couple bugs in List() where the wrong copy of an object was
  being used
- Add the "input" object to the list of existing objects so that it can
  be found in get, list, and update
- Implement Update() and expose and Update field on the test client so
  that callers can make assertions about the updates made

### CODE REVIEW

File: pkg/router/tester/tester.go [hunk 1]
```diff
@@ -143,7 +143,7 @@ func NewRequest(t *testing.T, scheme *runtime.Scheme, input kclient.Object, exis
 
 	return router.Request{
 		Client: &Client{
-			Objects:   existing,
+			Objects:   append(existing, input),
```
Comment (cjellick) [inline pkg/router/tester/tester.go:146]: do input.deepCopy()

File: pkg/router/tester/tester.go [hunk 2]
```diff
@@ -141,9 +141,16 @@ func NewRequest(t *testing.T, scheme *runtime.Scheme, input kclient.Object, exis
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	inputCopy := genericObj.(kclient.Object)
+	copy(inputCopy, input)
+
 	return router.Request{
 		Client: &Client{
-			Objects:   existing,
+			Objects:   append(existing, inputCopy),
```
Comment (ibuildthecloud) [inline pkg/router/tester/tester.go:153]: I think you just need `input.DeepCopyObject().(kclient.Object)`  You don't need all the gvk, copy stuff do you?

Comment (cjellick) [reply pkg/router/tester/tester.go:153]: yeah, that works

**Comment by cjellick** [comment]
@ibuildthecloud ready for re-review

**Pull request merged**